driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java

```diff
@@ -638,4 +720,310 @@ public Prune shouldPrune(final UsageTrackingInternalConnection usageTrackingConn
+                    connection.closeSilently();
+                }
+            } else {//acquired a permit, phase two
+                connectionCreated(//a connection is considered created only when it is ready to be open
+                        connectionPoolListener, getId(connection));
+                return openAndReleasePermit(connection, asyncCallback);
+            }
+        }
+
+        /**
+         * @return Either {@code null} if a permit has been acquired, or a {@link PooledConnection}
+         * if {@code tryGetAvailable} is {@code true} and an {@linkplain PooledConnection#opened() opened} one becomes available while
+         * waiting for a permit.
+         * @throws MongoTimeoutException If timed out.
+         */
+        @Nullable
+        private PooledConnection acquirePermitOrGetAvailableOpenConnection(final boolean tryGetAvailable, final Timeout timeout)
+                throws MongoTimeoutException {
+            PooledConnection availableConnection = null;
+            tryLock(timeout);
+            try {
+                if (tryGetAvailable) {
+                    addLastDesiredConnectionSlot();
+                }
+                long remainingNanos = timeout.remainingNanosOrInfinite();
+                while (permits == 0) {
+                    availableConnection = tryGetAvailable ? tryGetAndRemoveFirstDesiredConnectionSlot() : null;
+                    if (availableConnection != null) {
+                        break;
+                    }
+                    if (Timeout.expired(remainingNanos)) {
+                        throw createTimeoutException(timeout);
+                    }
+                    remainingNanos = awaitNanos(remainingNanos);
+                }
+                if (availableConnection == null) {
+                    assert permits > 0 : permits;
+                    permits--;
+                } else {
+                    assert availableConnection.opened();
+                    if (LOGGER.isTraceEnabled()) {
+                        LOGGER.trace(format("Received opened connection [%s] to server %s",
+                                getId(availableConnection), serverId.getAddress()));
+                    }
+                }
+                return availableConnection;
+            } finally {
+                try {
+                    if (tryGetAvailable && availableConnection == null) {//the desired connection slot has not yet been removed
+                        removeLastDesiredConnectionSlot();
+                    }
+                } finally {
+                    lock.unlock();
+                }
+            }
+        }
+
+        private void releasePermit() {
+            lock();
+            try {
+                assert permits < maxPermits : permits;
+                permits++;
+                condition.signal();
+            } finally {
+                lock.unlock();
+            }
+        }
+
+        private void addLastDesiredConnectionSlot() {
```

Comment:
Suggestion: Describe what the intent is? That way the logic becomes more obvious of what this method is used for. eg:  `signalDemandForConnection()`

Reply:
The methods are now called `expressDesireToGetAvailableConnection`, `tryGetAvailableConnection`, `giveUpOnTryingToGetAvailableConnection`. Cannot use "signal" because that will cause confusion related to the method `condition.signal`, which is not called by any of these methods.

Done.

Reply:
Thanks